diff --git a/zig/private/cc_helper.bzl b/zig/private/cc_helper.bzl
index 910bd26..e65a3a2 100644
--- a/zig/private/cc_helper.bzl
+++ b/zig/private/cc_helper.bzl
@@ -6,7 +6,7 @@ load("@rules_cc//cc/common:cc_common.bzl", "cc_common")
 def need_translate_c(cc_info):
     return cc_info.compilation_context and (cc_info.compilation_context.headers or cc_info.compilation_context.defines)
 
-def find_cc_toolchain(ctx, *, mandatory, features = [], disabled_features = []):
+def find_cc_toolchain(ctx, *, mandatory):
     """Extracts a CcToolchain from the current target's context
 
     Args:
@@ -23,7 +23,7 @@ def find_cc_toolchain(ctx, *, mandatory, features = [], disabled_features = []):
     feature_configuration = cc_common.configure_features(
         ctx = ctx,
         cc_toolchain = cc_toolchain,
-        requested_features = ctx.features + features,
-        unsupported_features = ctx.disabled_features + disabled_features,
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
     )
     return cc_toolchain, feature_configuration
diff --git a/zig/private/common/translate_c.bzl b/zig/private/common/translate_c.bzl
index 00b1046..1970387 100644
--- a/zig/private/common/translate_c.bzl
+++ b/zig/private/common/translate_c.bzl
@@ -79,11 +79,7 @@ def zig_translate_c(*, ctx, name, canonical_name, zigtoolchaininfo, global_args,
     # If there is a CC toolchain, add builtin directories.
     # This allows including to extra headers provided directly by the toolchain.
     # E.g. <os/log.h> on macOS.
-    cc_toolchain, cc_feature_configuration = find_cc_toolchain(
-        ctx,
-        mandatory = False,
-        disabled_features = ["thin_lto"],
-    )
+    cc_toolchain, cc_feature_configuration = find_cc_toolchain(ctx, mandatory = False)
 
     # Detect if the toolchain is the local apple cc toolchain since it requires
     # special handling to get the builtin headers included.
@@ -121,6 +117,8 @@ def zig_translate_c(*, ctx, name, canonical_name, zigtoolchaininfo, global_args,
         "-nobuiltininc",
         "-fmodule-libs",
         "-D__building_module(x)=0",
+        "-D__zig__=1",
+        "-D_ZIG_TRANSLATE_C=1",
     ])
 
     if cc_toolchain:
@@ -199,9 +197,9 @@ def zig_translate_c(*, ctx, name, canonical_name, zigtoolchaininfo, global_args,
         progress_message = "zig translate-c %{label}",
         execution_requirements = {tag: "" for tag in ctx.attr.tags},
         env = {
-            "ZIG_GLOBAL_CACHE_DIR": zigtoolchaininfo.zig_cache,
+            "ZIG_GLOBAL_CACHE_DIR": "/tmp/zig-working-cache",
             "ZIG_LIB_DIR": zigtoolchaininfo.zig_lib_path,
-            "ZIG_LOCAL_CACHE_DIR": zigtoolchaininfo.zig_cache,
+            "ZIG_LOCAL_CACHE_DIR": "/tmp/zig-working-cache",
         },
         tools = zigtoolchaininfo.zig_files,
         toolchain = "//zig:toolchain_type",
diff --git a/zig/private/common/zig_build.bzl b/zig/private/common/zig_build.bzl
index 92b20fe..bb7e4c7 100644
--- a/zig/private/common/zig_build.bzl
+++ b/zig/private/common/zig_build.bzl
@@ -2,9 +2,9 @@
 
 load("@bazel_skylib//lib:paths.bzl", "paths")
 load("@build_bazel_rules_android//:cc_common_link.bzl", "cc_common_link")
-load("@rules_cc//cc/common:cc_helper.bzl", "cc_helper")
-load("@rules_cc//cc:defs.bzl", "cc_common", "CcInfo")
 load("@rules_cc//cc:find_cc_toolchain.bzl", "use_cc_toolchain")
+load("@rules_cc//cc/common:cc_common.bzl", "cc_common")
+load("@rules_cc//cc/common:cc_info.bzl", "CcInfo")
 load("//zig/private:cc_helper.bzl", "find_cc_toolchain", "need_translate_c")
 load(
     "//zig/private/common:bazel_builtin.bzl",
@@ -472,6 +472,11 @@ buildozer 'move cdeps deps *' {target}
         execution_requirements = {tag: "" for tag in ctx.attr.tags},
         tools = zigtoolchaininfo.zig_files,
         toolchain = "//zig:toolchain_type",
+        env = {
+            "ZIG_GLOBAL_CACHE_DIR": "/tmp/zig-working-cache",
+            "ZIG_LOCAL_CACHE_DIR": "/tmp/zig-working-cache",
+            "HOME": "/tmp/zig-working-cache",
+        },
     )
 
     linkopts = location_expansion(
@@ -708,9 +713,6 @@ buildozer 'move cdeps deps *' {target}
     else:
         fail("Unknown rule kind '{}'.".format(kind))
 
-    if root_module.cc_info:
-        transitive_data.append(depset(cc_helper.get_dynamic_libraries_for_runtime(root_module.cc_info.linking_context, True)))
-
     providers.extend([
         DefaultInfo(
             executable = default_output if default_output_is_executable else None,
diff --git a/zig/private/common/zig_cache.bzl b/zig/private/common/zig_cache.bzl
index 526884f..7a86464 100644
--- a/zig/private/common/zig_cache.bzl
+++ b/zig/private/common/zig_cache.bzl
@@ -41,5 +41,5 @@ def zig_cache_output(*, zigtoolchaininfo, args):
       zigtoolchaininfo: ZigToolchainInfo.
       args: Args; mutable, Append the Zig cache flags to this object.
     """
-    args.add_all(["--cache-dir", zigtoolchaininfo.zig_cache])
-    args.add_all(["--global-cache-dir", zigtoolchaininfo.zig_cache])
+    args.add_all(["--cache-dir", "/tmp/zig-working-cache"])
+    args.add_all(["--global-cache-dir", "/tmp/zig-working-cache"])
