diff --git a/zml/module.zig b/zml/module.zig
index c65017a..1421cd9 100644
--- a/zml/module.zig
+++ b/zml/module.zig
@@ -366,7 +366,7 @@ fn collectInputInfo(allocator: std.mem.Allocator, partitioning: Partitioning, v:
     };
 }
 
-const EmitMlirResult = struct {
+pub const EmitMlirResult = struct {
     func: *mlir.Operation,
     input_info: InputInfo,
     output_info: OutputInfo,
@@ -380,7 +380,7 @@ fn finalizeAttributeList(allocator_: std.mem.Allocator, mlir_ctx: *mlir.Context,
     return res;
 }
 
-fn emitMlir(compilation_context: *CompilationContext, comptime func: anytype, args: stdx.meta.FnArgs(func)) !EmitMlirResult {
+pub fn emitMlir(compilation_context: *CompilationContext, comptime func: anytype, args: stdx.meta.FnArgs(func)) !EmitMlirResult {
     var arena = std.heap.ArenaAllocator.init(compilation_context.allocator);
     defer arena.deinit();
 
@@ -510,7 +510,7 @@ fn setXlaOverrideFlag(map: *c.upb_Map, flag: []const u8, value: anytype, upb_are
     }
 }
 
-fn compileModuleToPjrtExecutable(arena: std.mem.Allocator, io: std.Io, platform: *const Platform, module: *const mlir.Module, partitioning: Partitioning, opts: CompilationOptions) !*pjrt.LoadedExecutable {
+pub fn compileModuleToPjrtExecutable(arena: std.mem.Allocator, io: std.Io, platform: *const Platform, module: *const mlir.Module, partitioning: Partitioning, opts: CompilationOptions) !*pjrt.LoadedExecutable {
     var upb_alloc: upb.Allocator = .init(arena);
     const upb_arena = c.upb_Arena_Init(null, 0, upb_alloc.inner());
     defer c.upb_Arena_Free(upb_arena);
diff --git a/zml/tensor.zig b/zml/tensor.zig
index edbadf8..fe11d30 100644
--- a/zml/tensor.zig
+++ b/zml/tensor.zig
@@ -63,6 +63,9 @@ pub const Tensor = struct {
         return self._shape.byteSize();
     }
 
+    pub var global_tracer: ?*std.ArrayListUnmanaged(Tensor) = null;
+    pub var global_tracer_allocator: ?std.mem.Allocator = null;
+
     /// Internal use
     ///
     /// Creates a tensor from a Shape and an mlir.Value.
@@ -79,6 +82,10 @@ pub const Tensor = struct {
             stdx.debug.internalAssert(sh.eql(other._shape), "Created a {f} from Mlir value but expected {f}", .{ other._shape, res._shape });
         }
 
+        if (global_tracer) |tracer| {
+            tracer.append(global_tracer_allocator.?, res) catch @panic("OOM in global trace");
+        }
+
         return res;
     }
 
